skimage/metrics/overlap.py

```diff
@@ -0,0 +1,246 @@
+"""
+Module defining a number of functions to quantify the overlap between shapes.
+
+For instance BoundingBoxes used for object-detection.
+"""
+import numpy as np
+
+
+class BoundingBox:
+    """
+    Construct an axis-aligned Bounding-Box
+    consisting of top left and bottom right corners.
+
+    The contructor uses the (r,c,..) coordinates for the top left corner and
+    either the coordinates of the botton right corner or the BoundingBox
+    dimensions (height, width,...).
+
+    BoundingBoxes can have 2 (rectangle), 3 (3D BoundingBox) or more dimensions.
+
+    Parameters
+    ----------
+    top_left : array-like of ints or floats
+        (r,c)-coordinates for the top left corner of the BoundingBox.
+
+    bottom_right : array-like of ints or floats, optional
+        (r,c)-coordinates for the bottom right corner of the BoundingBox.
+
+    dimensions : array-like of ints or floats, optional
+        dimensions of the BoundingBox (height, width). The default is None.
+
+    Raises
+    ------
+    ValueError
+        If neither or both of bottom_right and dimensions are provided.
+
+    Attributes
+    ----------
+    top_left : array of int or float
+        The top left corner of the BoundingBox.
+    bottom_right : array of int or float
+        The bottom right corner of the BoundingBox.
+    """
+
+    def __init__(self, top_left, *, bottom_right=None, dimensions=None):
+        self.top_left = np.asarray(top_left)
+
+        if (bottom_right is None) and (dimensions is None):
+            raise ValueError("Specify one of bottom_right or dimensions.")
+
+        if (bottom_right is not None) and (dimensions is not None):
+            raise ValueError("Specify bottom_right or dimensions, not both.")
+
+        if bottom_right is not None:
+            if not np.all(top_left <= bottom_right):
+                raise ValueError("Bottom right corner should have coordinates "
+                                 "larger or equal to the top left corner.")
+            self.bottom_right = np.asarray(bottom_right)
+
+        elif dimensions is not None:
+            dimensions = np.asarray(dimensions)
+            if not (dimensions >= 0).all():
+                raise ValueError("Dimensions should be positive.")
+            self.bottom_right = self.top_left + dimensions
+
+    @property
+    def height(self):
+        # use negative indexing in anticipation of nD hyperBoundingBoxes.
+        return self.bottom_right[-2] - self.top_left[-2]
+
+    @property
+    def width(self):
+        # use negative indexing in anticipation of nD hyperBoundingBoxes.
+        return self.bottom_right[-1] - self.top_left[-1]
+
+    @property
+    def ndim(self):
+        return len(self.top_left)
+
+    def __bool__(self):
+        return True
+
+    def __eq__(self, other: BoundingBox):
```

Comment:
I also have test failing locally after approving the reviewed changes 😅
Somehow the BoundingBox type is not recognized. Is it something to change in the package __init__ ?

Reply:
Hum... I thin k this is due to Postponed Evaluation of Annotations ([PEP 563](https://www.python.org/dev/peps/pep-0563/))...
You are probably using python 3.7 right?
We must add in the imports
```
import sys
if sys.version_info.major == 3 and sys.version_info.minor ==7:
    from __future__ import annotations
```
Alternatively, we may remove the type annotations :slightly_smiling_face:

Reply:
@rfezzani you don't need to import sys. Just `from __future__ import annotations`. Which was in there a few commits ago. ;)

Reply:
I proposed this to avoid warnings described in the [first draft](https://mail.python.org/pipermail/python-ideas/2017-September/047031.html) of the PEP:

> Deprecation policy
> ------------------
> 
> In Python 3.7, a ``__future__`` import is required to use the described
> functionality and a ``PendingDeprecationWarning`` is raised by the
> compiler in the presence of type annotations in modules without the
> ``__future__`` import.  In Python 3.8 the warning becomes a
> ``DeprecationWarning``.  In the next version this will become the
> default behavior.

But, you are right, this seems not necessary in the last version of the PEP :).
But again, type hints are not yet mandatory in `skimage`, so we can avoid it :).

Reply:
In this case, I annotated it because I think it is a nice documentation when reading the function, ie we won't compare with a 4-tuple, or whatever else someone might want to do 😂

Reply:
I reverted the commit to keep the from future import statement for the moment.
We can always optimize it later.